File: OJs/MDSWIN2.cpp
// #include <iostream>
// #include <vector>
// #include <math.h>
// #include <string.h>
// #include <algorithm>
// #include <unordered_map>
#include<bits/stdc++.h>

using namespace std;

long MOD = 998244353;
const long MAX = 100001;
const long MAXQ = 10001;
long facts[MAX] ,times[MAX], num[MAX], cnt[MAX], all ,belong[MAX] ,nums[MAX], comb[MAXQ][MAXQ];
bool vis[MAX];

struct Instruct {
    long l, r; 
    long id;
    friend bool operator < (const Instruct& a, const Instruct& b) {
        if(belong[a.l]==belong[b.l]) return a.r < b.r;
        return belong[a.l] < belong[b.l];
    }
} instruct[MAX];

void fact_init() {
    facts[0] = 1;
    for(int i=1;i<MAX;i++) facts[i] = (facts[i-1]*i)%MOD;
}

void combinations(long _size){
    for(int i=0;i<_size;i++) {
        comb[0][i] = 0; comb[i][0] = 1;
    }
    for(int i=1;i<_size;i++) for(int j=1;j<=i;j++) {
        comb[i][j] = (comb[i-1][j]+comb[i-1][j-1])%MOD;
    }
}

long gcd(long a,long b){
    if(b==0) return a;
    return gcd(b,a%b);
 }

void eucl_gcd(long a,long b,long &x,long &y) {
    if(!b) {
        x = 1, y=0;
        return;
    }
    eucl_gcd(b,a%b,x,y);
    long l=x;
    x=y;
    y=l-a/b*y;
}

void _add(long x) {
    ++num[++cnt[x]];
    times[++all]=cnt[x];
    --num[cnt[x]-1];
}

void _del(long x) {
    ++num[--cnt[x]];
    times[++all]=cnt[x];
    --num[cnt[x]+1];
}

long select_ish(long MOD, long n,long m) {
    if(n<MAXQ) return comb[n][m];
    if(m==1 || m==n-1) return n;
    if(m>n) return 0;
    if(n==m) return 1;
    long nn=facts[n], mm=(facts[m]*facts[n-m])%MOD;
    long d=gcd(nn,mm);
    nn/=d, mm/=d;
    long x,y;
    eucl_gcd(mm,MOD,x,y);
    x=(x+MOD)%MOD;
    return (x*nn)%MOD;
}

long get_valof() {
	int _ret;
    long k = 0;
    long sum = 0;
    long out = 0; 
    long ans, _tmp;
    vector<long> vec;
    for(long j=1;j <= all;j++) if(!vis[times[j]]&&num[times[j]]>0) {
        vec.push_back(times[j]);
        times[++k]=times[j];
        vis[times[j]]=true;
    }
    all=k;
    for(int j=1;j<=k;j++) vis[times[j]] = false;
    for(long i: vec) if(num[i]%2) sum ^= i;
    if(sum==0) return 0;
    for(long i: vec) if(num[i]>0) {
        ans = sum^i;
        if(ans > i) continue;
        _tmp = select_ish(MOD, i, i-ans);
        _tmp = (_tmp*num[i])%MOD;
        out = (out+_tmp)%MOD;
    }
    _ret = out - sum;
    return out;
}

void rescro(long m) {
    long L = 1, R = 0 ,M;
    vector<long> _val(m+1);
    vector<bool> size_of(m/2);
    for(int i=1;i<=m;i++) {
        for(;R<instruct[i].r;R++) _add(nums[R+1]);
        for(;L>instruct[i].l;L--) _add(nums[L-1]);
        for(;R>instruct[i].r;R--) _del(nums[R]);
        for(;L<instruct[i].l;L++) _del(nums[L]);
        _val[instruct[i].id] = get_valof();
    }
    for(int i=1;i<=m;i++) printf("%ld\n", _val[i]);
}

void solve(){
    long  n, q, l, r;
    long ret = -1;
    unordered_map<long, long> dict_ll;
    scanf("%ld", &n);
    long sq = sqrt(n);
    for(int i=1;i<=n;i++) {
        scanf("%ld", &nums[i]); 
        belong[i] = (i-1)/sq+1;
    }
    for(int i=1;i<=n;i++) {
        if(dict_ll.find(nums[i])==dict_ll.end()) dict_ll[nums[i]] = dict_ll.size();
        nums[i] = dict_ll[nums[i]];
    }
    scanf("%ld", &q);
    for(int i=1;i<=q;i++) {
        scanf("%ld%ld", &instruct[i].l, &instruct[i].r);
        instruct[i].id = i;
    }
    sort(instruct+1, instruct+q+1);
    rescro(q);
}

void _clear(int _a){
	memset(times, _a, sizeof(times));
    memset(num, _a, sizeof(num));
    memset(cnt, _a, sizeof(cnt));
    memset(vis, _a, sizeof(vis));
    all = _a;
}

void cases(){
	_clear(0);
    solve();
}

int main(){
    fact_init();
    combinations(MAXQ); // here
    int t; cin>>t;
    while(t--) cases();
    
}

File: OJs/MDSWIN2.py
from math import sqrt
mod = 998244353
maxn = 100001

nums = [0] # maxn
midn = 1001

belong = [0] #maxn
times = [0] #maxn
num = [0] #maxn
cnt = [0] #maxn 
ALL = 0 
vis = [0] #maxn

facts = [1] # maxn	
for i in range(1 ,maxn):facts.append((facts[i-1]*i)%mod)

comb = [[None]*midn for _ in range(midn)]
for i in range(midn):
	comb[0][i] ,comb[i][0] = 0 ,1
for i in range(1 ,midn):
	for j in range(1 ,i+1):
		comb[i][j] = (bool(comb[i-1][j])+bool(comb[i-1][j-1]))%mod


class CMD(object):
	def __init__(self ,l ,r ,_id):
		self.l = l
		self.r = r
		self.id = _id

	def __lt__(self, other):
		if belong[self.l] == belong[other.l]:
			return self.r < other.r
		return belong[self.l] < belong[other.l]

cmd = [None]
def gcd(a ,b):
	if not b : return a
	return gcd(b ,a%b)

def e_gcd(a ,b ,x ,y):
	if not b:
		x ,y = 1 ,0
		return (x ,y)
	x ,y = e_gcd(b ,a%b ,x ,y)
	l = x
	x = y
	y = l-a/b*y
	# x ,y = y ,x-a/b*y
	return (x ,y)

def choose(mod ,n ,m):
	if n<midn :return comb[n][m]
	if m == 1 or m == n-1 : return n
	if m>n : return 0
	if n ==m : return 1
	nn = facts[n]
	mm = (facts[m]*facts[n-m])%mod
	d = gcd(nn ,mm)
	nn /= d
	mm /= d
	x ,y = e_gcd(mm ,mod ,0 ,0) # here
	x = (x+mod)%mod
	return (x*nn)%mod

def _add(x):
	global ALL
	cnt[x] += 1
	num[cnt[x]] += 1
	ALL += 1
	times[ ALL] = cnt[x]
	num[cnt[x]-1] -= 1

def _del(x):
	global ALL
	cnt[x] -= 1
	num[cnt[x]] += 1
	ALL += 1
	times[ALL] = cnt[x]
	num[cnt[x] + 1] -= 1

def cal():
	global ALL
	k = 0
	vec = []
	for j in range(1 ,ALL +1):
		if not vis[times[j]] and num[times[j]]>0 :
			vec.append(times[j])
			k += 1
			times[k] = times[j]
			vis[times[j]] = True

	ALL = k
	for j in range(1 ,k+1): 
		vis[times[j]] = False
	_sum = 0
	for i in vec:
		if num[i]%2 :
			_sum ^= i
	if _sum == 0 :
		return 0
	_res = 0
	ans = 0
	tmp = 0
	for i in vec:
		if num[i] > 0:
			ans = _sum ^i
			if ans > i :
				continue
			tmp = choose(mod ,i ,i-ans)
			tmp = (tmp*num[i])%mod
			_res = (_res+tmp)%mod
	return _res

def solve(m):
	L = 1
	R = 0
	res = [0 for i in range(m+1)]
	for i in range(1 ,m+1):
		while R<cmd[i].r :
			_add(nums[R+1])
			R += 1
		while L>cmd[i].l :
			_add(nums[L-1])
			L -= 1
		while R>cmd[i].r :
			_del(nums[R])
			R -= 1
		while L<cmd[i].l:
			_del(nums[L])
			L += 1

		res[cmd[i].id] = cal()
	for i in range(1 ,m+1):
		print(res[i])

def calc():
	# global cmd
	n = int(input())
	sq = sqrt(n)
	nums[0] = 0
	belong[0] = 0
	nums.extend(list(map(int , input().split())))
	for i in range(1 ,n+1):
		# nums.append(*map(int , input().split()))
		belong.append((i-1)/sq+1)

	um = dict()

	for i in range(1 ,n+1):
		if not (nums[i] in um.keys()):
			um[nums[i]] = len(um)
		nums[i] = um[nums[i]]

	q = int(input())
	cmd[0] = CMD(0 ,0 ,0)
	for i in range(1 ,q+1):
		cmd.append(CMD(*map(int , input().split()) ,i))
		# cmd[i].id = i
	for i in range(maxn):
		times.append(0)
		num.append(0)
		cnt.append(0)
		vis.append(0)
	ALL = 0
	cmd.sort()
	# cmd = [-1] + cmd
	solve(q)


def main():
	# init()
	for i in range(int(input())):
		calc()

if __name__ == '__main__':
	main()